app/com/linkedin/drelephant/analysis/genericheuristics/GenericPercentileHeuristic.java

```diff
@@ -0,0 +1,231 @@
+    if (severityPercentileCutoffs != null) {
+      _severityPercentileCutoffs = severityPercentileCutoffs;
+    }
+    logger.info(heuristicName + " will use " + PERCENTILE_CUTOFFS_SEVERITY + " with the following threshold settings: "
+            + Arrays.toString(_severityPercentileCutoffs));
+
+
+    String percentileUpdateDelayString = paramMap.get(PERCENTILE_UPDATE_DELAY);
+    if (percentileUpdateDelayString != null) {
+      _percentileUpdateDelay = Long.valueOf(percentileUpdateDelayString);
+    }
+    logger.info(heuristicName + " will use " + PERCENTILE_UPDATE_DELAY + " with the following threshold settings: "
+            + _percentileUpdateDelay);
+  }
+
+  @Override
+  public HeuristicConfigurationData getHeuristicConfData() {
+    return _heuristicConfData;
+  }
+
+  private Severity getSeverity(long value) {
+    return Severity.getSeverityAscending(
+            value,
+            _severityAbsoluteCutoffs[0],
+            _severityAbsoluteCutoffs[1],
+            _severityAbsoluteCutoffs[2],
+            _severityAbsoluteCutoffs[3]);
+  }
+
+  @Override
+  public HeuristicResult apply(T data) {
+    if (!shouldApply(data))
+      return null;
+    updateCutoffs();
+
+    long value;
+    ArrayList<HeuristicResultDetails> details = new ArrayList<HeuristicResultDetails>();
+    try {
+      value = getValue(data, details);
+    } catch (RuntimeException e) {
+      logger.info(String.format("%s failed to get value: %s",
+              _heuristicConfData.getHeuristicName(),
+              e.getMessage()));
+      return null;
+    }
+
+    int score;
+    try {
+      score = getScore(data, value);
+    } catch (RuntimeException e) {
+      logger.info(String.format("%s failed to get score: %s",
+              _heuristicConfData.getHeuristicName(),
+              e.getMessage()));
+      return null;
+    }
+
+
+    HeuristicResult result = new HeuristicResult(_heuristicConfData.getClassName(),
+            _heuristicConfData.getHeuristicName(),
+            getSeverity(value),
+            score,
+            details);
+    result.addResultDetail(getPercentileHeuristicDetailName(), format(value), Long.toString(value));
+    result.addResultDetail(percentileInfo(_severityPercentileCutoffs[0], "low"), format(_severityAbsoluteCutoffs[0]));
+    result.addResultDetail(percentileInfo(_severityPercentileCutoffs[1], "moderate"), format(_severityAbsoluteCutoffs[1]));
+    result.addResultDetail(percentileInfo(_severityPercentileCutoffs[2], "severe"), format(_severityAbsoluteCutoffs[2]));
+    result.addResultDetail(percentileInfo(_severityPercentileCutoffs[3], "critical"), format(_severityAbsoluteCutoffs[3]));
+
+    return result;
+  }
+
+  protected static String percentileInfo(double percentile, String severity) {
+    return String.format("Current %s%% (over last %s days) (%s severity)",
+            percentile, DAYS_TO_USE_FOR_PERCENTILE, severity);
+  }
+
+  private void updateCutoffs() {
```

Comment:
I am not sure I understand what this function does. Can you add some comments to explain where and how it's used ? Also, none of the other heuristics read from DB to decide on the score. It seems that this one is doing that. Can you explain why that is needed ?